Source/JavaScriptCore/interpreter/Interpreter.cpp

```diff
@@ -124,12 +124,21 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
 
     JSValue program = callFrame->argument(0);
     JSString* programString = nullptr;
+    bool isTrusted = false;
     if (LIKELY(program.isString()))
         programString = asString(program);
-    else if (Options::useTrustedTypes()) {
-        auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
-        if (!code.isNull())
-            programString = jsString(vm, code);
+    else if (Options::useTrustedTypes() && program.isObject()) {
+        auto* structure = globalObject->trustedScriptStructure();
+        if (structure == program.structureOrNull()) {
+            programString = program.toStringOrNull(globalObject);
+            isTrusted = true;
+        } else {
+            auto code = globalObject->globalObjectMethodTable()->codeForEval(globalObject, program);
+            if (!code.isNull()) {
+                programString = jsString(vm, code);
```

Comment:
Why do we need to wrap it with string? This is unnecessary allocation.

Reply:
Because it's just a `String` wrapping in jsString because programString is a JSString?

Reply:
I mean, there is very high possibility that this input `code` is not changed as a result. So we should avoid allocating new string here for the major path. We normally do not care about this, but `eval` etc. is hot path, and very small changes easily affect on the real world performance.

Reply:
What would you recommend we do instead?

Reply:
@Constellation just coming back to this and I've realised this isn't the hot path. The structure check is the hot path. This is for when they're evaling non trusted objects OR where something unusual has happened to the trusted object such that the structure match fails.

Given that I don't think we have to worry about these allocations as much? Unless you're saying that eval of generic objects is a hotpath?

If this does indeed require code changes, I'm also unsure what you mean by `there is very high possibility that this input code is not changed as a result` code is a String and we need a JSString at this point in the code flow? Do you have a code snippet example of what you'd like this to look like?

Reply:
Why do we need to allocate this string? I think you can just get `String programSource` early, changing `reportViolationForUnsafeEval` to take `const String&` instead of `JSString*` is enough, right?

Reply:
I wanted to avoid doing too much of a refactor unrelated to the code I'm trying to add. But I've updated the code as suggested.